### DISCUSSION TITLE
EMD distance profile

### DISCUSSION THREAD

**Discussion by testereng** [matrix-profile-foundation/matrixprofiler#12] [Ideas]
Hello members
   not sure if it is the right place (might be in tsmp depository), but I found that the Earth Mover's Distance (or Wasserstein) could provide good results in matching sequences independantly of the order (a little like MPdist).
 To do so I am creating a  vector of  quantiles at each time of the time series, and the EMD will be simply the euclidean distance between this vector and the quantile vector of the subsequence you want to match.
 I got a (scrappy) piece of code below taking advantage of sliding fast quantile update (using 'caTools'), and 'Rfast' to get (brutal) distance between my reference and the other subsequences

```R
emd_slid <- function(ts, ref, nbq) {
  pr <- seq(0, 1, by = nbq)
  qref <- quantile(x = ref, probs = pr)
  R <- caTools::runquantile(x = ts, k = length(ref),probs = pr)
  D <- Rfast::dista(xnew = t(qref), x = R, square = FALSE)
  return(D)
}
```

What do you think about it? Would there be any way to extend it (I find similarities in classic MP, since somehow there is 1 point change in sliding subsequence, so quantile is not recalcluated from scratch..)
Cheers!

**Comment by franzbischoff** [comment]
If this is a distance measurement, this can be a `matrixprofiler` feature.
Indeed, the ideal would be in the future we have something that could accept any algorithm that outputs a distance and then outputs the distance profile.

Just a thought for now.
